### ISSUE TITLE
handle overlapping `nlabels`

### ISSUE THREAD

**Issue by SimonEnsemble** [JuliaPlots/GraphMakie.jl#131]
[feature request]

I have a graph with many nodes. can't show all node labels with `nlabels` without some ugly overlap. 

but Gadfly does it nicely with their `Geom.label`. would be awesome if `GraphMakie.jl` could do this. their code for this is [here](https://github.com/GiovineItalia/Gadfly.jl/blob/master/src/geom/label.jl). showing `GraphMakie` vs `Gadfly` below.

<img width="539" alt="Screenshot 2023-05-09 at 9 31 29 AM" src="https://github.com/MakieOrg/GraphMakie.jl/assets/30643894/dcb537a2-b277-4739-aeee-c83802c6ae8d">
<img width="539" alt="Screenshot 2023-05-09 at 9 31 18 AM" src="https://github.com/MakieOrg/GraphMakie.jl/assets/30643894/3c506c9f-f05e-4ed3-a9fb-cb7780b1f5cc">

**Comment by asinghvi17** [comment]
This would probably also involve an improvement on Makie's end.  I tried to implement a `text_repel!` recipe in Makie from some R code, but having the Julia code from https://github.com/GiovineItalia/Gadfly.jl/blob/master/src/geom/label.jl helps quite a bit.  Will see if I can do something about it, but if anyone is interested it would be a pretty useful feature!

**Comment by SimonEnsemble** [comment]
oh, I pre-computed the layout from `NetworkLayout`, then manually scaled the coords of the blue points, to draw this in Gadfly. I am just screen-shotting part of the visualization. (didn't want to put pictures of my unpublished research online haha).

@asinghvi17 yes, would be very useful for `Makie` in general, when labeling points of a scatter plot! we need this for a scatter plot on another project (`text!` would lead to crazy-overlapping labels yet we want to label as many as possible.).

if only the text were spheres; then `NetworkLayout` could be used to repel the text away from each other and the scatter plot points.

**Comment by hdavid16** [comment]
I'd expect you would not get as ugly of an overlap if the nodes were distributed as in your gadfly plot.

**Comment by hexaeder** [comment]
> This would probably also involve an improvement on Makie's end. 

You mean a new alignment option `text` recipe, such that `text(positions; text=[...], align=:dynamic)` would choose an appropriate anchor for each textbox with little overlap? How'd one handle the zoom level? Normally, text happens in pixelspace while the positions happen to be in markerspace...

**Comment by asinghvi17** [comment]
Yep, something like that.  Text is laid out in screen space in any case, so it's not too difficult to do.  In terms of zoom level - we would have to recompute whenever axis limits change.

**Comment by SimonAB** [comment]
I opened a PR that adds opt-in `nlabels_auto_align` for this: https://github.com/MakieOrg/GraphMakie.jl/pull/257

It places each label in the largest angular gap among incident edges (default remains off for compatibility).
